examples/tic-tac-toe/cli/src/client.rs

```diff
@@ -0,0 +1,749 @@
+
+        let gas_price = self
+            .wallet
+            .get_reference_gas_price()
+            .await
+            .context("Error fetching reference gas price")?;
+
+        let tx_kind = TransactionKind::ProgrammableTransaction(tx);
+
+        // Gas Estimation
+        let tx_data = client
+            .transaction_builder()
+            .tx_data_for_dry_run(
+                sender,
+                tx_kind.clone(),
+                max_budget,
+                gas_price,
+                /* gas_payment */ None,
+                /* gas_sponsor */ None,
+            )
+            .await;
+
+        let DryRunTransactionBlockResponse { effects, .. } = client
+            .read_api()
+            .dry_run_transaction_block(tx_data.clone())
+            .await
+            .context("Error estimating gas budget")?;
+
+        let gas_used = effects.gas_cost_summary();
+        let overhead = 1000 * gas_price;
+        let net_used = gas_used.net_gas_usage();
+        let computation = gas_used.computation_cost;
+
+        let budget = overhead + (net_used.max(0) as u64).max(computation);
+
+        let payment = self
+            .select_coins(sponsor.unwrap_or(sender), budget, &tx_kind)
+            .await?;
+
+        Ok(if let Some(sponsor) = sponsor {
+            TransactionData::new_with_gas_coins_allow_sponsor(
+                tx_kind, sender, payment, budget, gas_price, sponsor,
+            )
+        } else {
+            TransactionData::new_with_gas_coins(tx_kind, sender, payment, budget, gas_price)
+        })
+    }
+
+    /// Find the max budget allowed for a transaction according to the current protocol config.
+    async fn max_gas_budget(&self) -> Result<u64> {
+        let client = self.client().await?;
+
+        let cfg = client.read_api().get_protocol_config(None).await?;
+        let Some(Some(SuiProtocolConfigValue::U64(max))) = cfg.attributes.get("max_tx_gas") else {
+            bail!("Couldn't find max gas budget");
+        };
+
+        Ok(*max)
+    }
+
+    /// Select Gas coins owned by `owner` to meet `balance`, avoiding input objects to the
+    /// transaction, `tx`.
+    async fn select_coins(
+        &self,
+        owner: SuiAddress,
+        balance: u64,
+        tx: &TransactionKind,
+    ) -> Result<Vec<ObjectRef>> {
+        let client = self.client().await?;
+        let coins = client.coin_read_api();
+        let exclude: Vec<_> = tx
+            .input_objects()?
+            .into_iter()
+            .filter_map(|input| match input {
+                InputObjectKind::ImmOrOwnedMoveObject((id, _, _)) => Some(id),
+                InputObjectKind::MovePackage(_) => None,
+                InputObjectKind::SharedMoveObject { .. } => None,
+            })
+            .collect();
+
+        Ok(coins
+            .select_coins(owner, None, balance as u128, exclude)
+            .await?
+            .into_iter()
+            .map(|coin| coin.object_ref())
+            .collect())
+    }
```

Comment:
I might be mistaking, but the wallet already provides this kind of functionality, so maybe you do not need to reimplement?
https://github.com/MystenLabs/sui/blob/main/crates/sui-sdk/src/wallet_context.rs#L180